Frameworks/Foundation/NSOperation.mm

```diff
@@ -131,139 +166,123 @@ - (void)setThreadPriority:(double)priority {
 }
 
 /**
  @Status Interoperable
 */
 - (void (^)(void))completionBlock {
-    return priv->completionBlock;
+    std::lock_guard<std::recursive_mutex> lock(_completionBlockLock);
+    return [[_completionBlock retain] autorelease];
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)isReady {
-    //  Note, check dependencies when we get them
-    int count = [priv->dependencies count];
-
-    for (int i = 0; i < count; i++) {
-        id op = [priv->dependencies objectAtIndex:i];
-
-        if (![op isFinished]) {
-            return NO;
-        }
-    }
-    return YES;
+- (BOOL)isCancelled {
+    std::lock_guard<std::recursive_mutex> lock(_finishLock);
+    return _cancelled;
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)isCancelled {
-    return priv->cancelled != 0;
+- (BOOL)isFinished {
+    return _finished;
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)isFinished {
-    return priv->finished != 0;
+- (BOOL)isExecuting {
+    return _executing;
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)isExecuting {
-    return priv->executing != 0;
+- (BOOL)isConcurrent {
+    return NO;
 }
 
 /**
  @Status Interoperable
 */
-- (void)start {
-    if (!priv->executing && !priv->finished) {
-        bool execute = false;
-
-        pthread_mutex_lock(&priv->finishLock);
-        if (!priv->cancelled) {
-            [self willChangeValueForKey:@"isExecuting"];
-            priv->executing = 1;
-            [self didChangeValueForKey:@"isExecuting"];
-            execute = true;
-        }
-        pthread_mutex_unlock(&priv->finishLock);
-        if (execute) {
-            [self main];
-        }
-
-        if (execute) {
-            [self willChangeValueForKey:@"isExecuting"];
-        }
-
-        [self _setFinished:YES
-            andPerformUnderLock:^{
-                if (execute) {
-                    priv->executing = 0;
-                    [self didChangeValueForKey:@"isExecuting"];
-                }
-            }];
-    }
+- (BOOL)isAsynchronous {
+    return NO;
 }
 
 /**
  @Status Interoperable
 */
-- (void)setFinished:(BOOL)finished {
-    [self _setFinished:finished andPerformUnderLock:nil];
-}
+- (void)start {
+    if (_finished) {
+        return;
+    }
 
-- (void)_setFinished:(BOOL)finished andPerformUnderLock:(void (^)())block {
-    // yes, this is ugly: priv->finished is an int though.
-    int newValue = finished ? 1 : 0;
-    pthread_mutex_lock(&priv->finishLock);
+    THROW_NS_IF(E_INVALIDARG, (_executing || ![self isReady]));
+    
+    _finishLock.lock();
+    BOOL shouldExecute = !_cancelled; // Note: in the cancelled case, [self main] is not called but isFinished will still be observable
+    if (shouldExecute) {
+        [self willChangeValueForKey:@"isExecuting"];
+        _executing = YES;
+        [self didChangeValueForKey:@"isExecuting"];
+    }
+    _finishLock.unlock();
+    
+    if (shouldExecute) {
+        NSAutoreleasePool* pool = [NSAutoreleasePool new];
+        [self main];
```

Comment:
NSOperation support being invoked "manually" - that is, outside of the Queue by user code. We still need an autorelease pool in that case.

---
In reply to: [69818876](https://github.com/Microsoft/WinObjC/pull/602#discussion_r69818876) [](ancestors = 69818876)